src/protocol/boolean/random_bits_generator.rs

```diff
@@ -0,0 +1,139 @@
+use super::solved_bits::{RandomBitsShare, SolvedBits};
+use crate::error::Error;
+use crate::ff::Field;
+use crate::protocol::context::SemiHonestContext;
+use crate::protocol::RecordId;
+use std::collections::HashMap;
+
+/// A struct that pre-generates and buffers random sharings of bits from the
+/// `SolvedBits` protocol. Any protocol who wish to use a random-bits can draw
+/// one by calling `take_one()`. It will call `SolvedBits` once the stock falls
+/// below `REFILL_THRESHOLD` until it fills up the empty slots.
+///
+/// Internally, it uses two pointers. `read_pointer` keeps track of the next
+/// item to return, and `write_pointer` points to the least recent empty slot.
+/// The pointers are `i8`. Incrementing them cause wrap around at the boundary,
+/// which makes the `buffer` hash map a ring buffer.
+#[allow(dead_code)]
+pub struct RandomBitsGenerator<'a, F: Field> {
+    context: SemiHonestContext<'a, F>,
+    buffer: HashMap<i8, RandomBitsShare<F>>,
+    counter: u32,
+    read_pointer: i8,
+    write_pointer: i8,
```

Comment:
Depends on where I place the generator instance, but most likely need to guard these with the mutex.